daemon/logger/awslogs/cloudwatchlogs.go

```diff
@@ -460,37 +411,30 @@ func (l *logStream) processEvent(events []wrappedEvent, unprocessedLine []byte,
 		}
 		line := unprocessedLine[:lineBytes]
 		unprocessedLine = unprocessedLine[lineBytes:]
-		if (len(events) >= maximumLogEventsPerPut) || (bytes+lineBytes+perEventBytes > maximumBytesPerPut) {
+		if (batch.Count() >= maximumLogEventsPerPut) || (batch.Size()+lineBytes+perEventBytes > maximumBytesPerPut) {
 			// Publish an existing batch if it's already over the maximum number of events or if adding this
 			// event would push it over the maximum number of total bytes.
-			l.publishBatch(events)
-			events = events[:0]
-			bytes = 0
+			l.publishBatch(batch)
 		}
-		events = append(events, wrappedEvent{
+		event := wrappedEvent{
 			inputLogEvent: &cloudwatchlogs.InputLogEvent{
 				Message:   aws.String(string(line)),
 				Timestamp: aws.Int64(timestamp),
 			},
-			insertOrder: len(events),
-		})
-		bytes += (lineBytes + perEventBytes)
+			insertOrder: batch.Count(),
+		}
+		batch.Add(event, lineBytes)
 	}
-	return events
 }
 
 // publishBatch calls PutLogEvents for a given set of InputLogEvents,
 // accounting for sequencing requirements (each request must reference the
 // sequence token returned by the previous request).
-func (l *logStream) publishBatch(events []wrappedEvent) {
-	if len(events) == 0 {
+func (l *logStream) publishBatch(batch *eventBatch) {
+	if batch.Count() == 0 {
```

Comment:
What if  `batch.Count` is 0, but `batch.bytes` is non-zero? Things could get out of whack. Probably better to `Reset` before return anyways.